src/qx/Simulator.cpp

```diff
@@ -13,67 +13,59 @@
 
 namespace qx {
 
-std::optional<SimulationError> Simulator::set(std::string const &filePath) {
-    auto analyzer = cqasm::v1x::default_analyzer("1.2");
-
-    auto analysisResult = analyzer.analyze(filePath);
-    if (!analysisResult.errors.empty()) {
-        program.reset();
-        std::stringstream error;
-        error << "Cannot parse and analyze file " << filePath << ":";
-
-        std::for_each(analysisResult.errors.begin(), analysisResult.errors.end(), [&error](auto const& s) { error << "\n" << s; });
-
-        return SimulationError{error.str()};
-    }
+using AnalysisResult = cqasm::v1x::analyzer::AnalysisResult;
+using Program = cqasm::v1x::ast::One<cqasm::v1x::semantic::Program>;
 
-    program = analysisResult.root;
-    assert(!program.empty());
-
-    if (!program->error_model.empty() &&
-        program->error_model->name != "depolarizing_channel") {
-        program.reset();
-        return SimulationError{"Unknown error model: " + program->error_model->name};
-    }
+namespace {
+AnalysisResult parseCQASMV1File(std::string const &filePath) {
+    auto analyzer = cqasm::v1x::default_analyzer("1.2");
 
-    return std::nullopt;
+    return analyzer.analyze(filePath);
 }
 
-std::optional<SimulationError> Simulator::setString(std::string const &s) {
+AnalysisResult parseCQASMV1String(std::string const &s) {
     auto analyzer = cqasm::v1x::default_analyzer("1.2");
 
-    auto analysisResult = analyzer.analyze_string(s);
+    return analyzer.analyze_string(s);
+}
+
+std::variant<Program, SimulationError> getProgramOrError(AnalysisResult const& analysisResult) {
     if (!analysisResult.errors.empty()) {
-        program.reset();
         std::stringstream error;
-        error << "Cannot parse and analyze string " << s << ": \n";
+        error << "Cannot parse and analyze cQASM: \n";
 
         std::for_each(analysisResult.errors.begin(), analysisResult.errors.end(), [&error](auto const& x) { error << x << "\n"; });
 
         return SimulationError{error.str()};
     }
 
-    program = analysisResult.root;
+    auto program = analysisResult.root;
     assert(!program.empty());
 
     if (!program->error_model.empty() &&
         program->error_model->name != "depolarizing_channel") {
         std::cerr << "Unknown error model: " << program->error_model->name
-                  << std::endl;
```

Comment:
I thought we had added `fmt` as a Conan dependency. In case you want to add it, you can take `libqasm` as a reference (one line in `conanfile.py`, and a few in `CMakeLists.txt`.
 
 It would turn this code into: `fmt::print(std::cerr, "Unknown error model: {}\n", program->error_model->name);`.

 It would need `#include <fmt/core.h>`, or `#include <fmt/ostream.h>`, not sure.

Reply:
All code in `Simulator.cpp` and `src/qx/Simulator.cpp` has been updated to use `fmt`.